### PR TITLE
Close #5: Add unoptimized minimax solver

### LINKED ISSUE #5: First approach to AI

### PR DESCRIPTION
**Pull request by ognick**

### CODE REVIEW

**Comment by fomkin** [comment]
Better naming for PR is "Close #5: Add unoptimized minimax solver"

File: src/main/scala/RunSolver.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import wop.game.{Point, TicTacToe, WopState}
+import wop.game.WopState.{Board, Draw, Foul, InProgress, Player, Select, SubBoard, Turn, Win}
+import wop.game.ai.WopSolver
+
+
+object RunSolver extends App {
+  def loop(state: WopState): Option[WopState] = {
+    println(state)
```
Comment (fomkin) [inline src/main/scala/RunSolver.scala]: Remove debug before making PR

File: src/main/scala/wop/game/ai/WopSolver.scala [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  val MAX_DEPTH = 2
+
+  def playerFactor(xo: WopState.XO): Int = xo match {
+    case WopState.Player.P1.xo => 1
+    case WopState.Player.P2.xo => -1
+    case WopState.XO.Empty => 0
+  }
+
+  def scan(subBoard: SubBoard)(implicit item: Item[WopState.XO]): Int = {
+    def loop(x: Int, y: Int, dx: Int, dy: Int, prev: Option[WopState.XO] = None, acc: Int = 0): Int = {
```
Comment (fomkin) [inline src/main/scala/wop/game/ai/WopSolver.scala]: Add `@tilerec` annotation

File: src/main/scala/wop/game/ai/WopSolver.scala [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+          case Some(p) if item.compare(curr, p) =>
+            loop(x + dx, y + dy, dx, dy, Some(curr), acc + playerFactor(curr) * 2)
+          case None => loop(x + dx, y + dy, dx, dy, Some(curr), acc + playerFactor(curr))
+          case _ => 0
+        } else loop(x + dx, y + dy, dx, dy, prev, acc)
+      }
+    }
+
+    val sizeRange = (0 until WopState.Size).toList
+    val vertical = (for (x: Int <- sizeRange) yield loop(x = x, y = 0, 0, 1)).sum
```
Comment (fomkin) [inline src/main/scala/wop/game/ai/WopSolver.scala]: You don't need `: Int`

File: src/main/scala/wop/game/ai/WopSolver.scala [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+object WopSolver {
+  val MAX_DEPTH = 2
+
+  def playerFactor(xo: WopState.XO): Int = xo match {
+    case WopState.Player.P1.xo => 1
+    case WopState.Player.P2.xo => -1
+    case WopState.XO.Empty => 0
+  }
+
+  def scan(subBoard: SubBoard)(implicit item: Item[WopState.XO]): Int = {
```
Comment (fomkin) [inline src/main/scala/wop/game/ai/WopSolver.scala:17]: What difference between `scan` and `TicTacToe#status`?

Comment (ognick) [reply src/main/scala/wop/game/ai/WopSolver.scala:17]: it is heuristic eval

Comment (fomkin) [reply src/main/scala/wop/game/ai/WopSolver.scala:17]: OK

**Comment by fomkin** [comment]
LGTM 👍

**Pull request merged**